sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1910,486 @@ def transform(self, X):
+                    replace=False, random_state=random_state)
+            else:
+                if self.ignore_implicit_zeros:
+                    column_data = np.zeros(shape=len(column_nnz_data),
+                                           dtype=X.dtype)
+                else:
+                    column_data = np.zeros(shape=n_samples, dtype=X.dtype)
+                column_data[:len(column_nnz_data)] = column_nnz_data
+
+            if not column_data.size:
+                # if no nnz, an error will be raised for computing the
+                # quantiles. Force the quantiles to be zeros.
+                self.quantiles_.append([0] * len(references))
+            else:
+                self.quantiles_.append(
+                    self._compute_quantile_one_column(column_data, references,
+                                                      random_state))
+        self.quantiles_ = np.transpose(self.quantiles_)
+
+    def fit(self, X, y=None):
+        """Compute the quantiles used for transforming.
+
+        Parameters
+        ----------
+        X : ndarray or sparse matrix, shape (n_samples, n_features)
+            The data used to scale along the features axis. If a sparse
+            matrix is provided, it will be converted into a sparse
+            ``csc_matrix``. Additionally, the sparse matrix needs to be
+            semi-positive if `ignore_implicit_zeros` is False.
+
+        Returns
+        -------
+        self : object
+            Returns self
+        """
+        X = check_array(X, accept_sparse='csc')
+        rng = check_random_state(self.random_state)
+
+        if self.n_quantiles <= 0:
+            raise ValueError("Invalid value for 'n_quantiles': %d. "
+                             "The number of quantiles must be at least one."
+                             % self.n_quantiles)
+
+        if self.subsample <= 0:
+            raise ValueError("Invalid value for 'subsample': %d. "
+                             "The number of subsamples must be at least one."
+                             % self.subsample)
+
+        if self.smoothing_noise is not None:
+            if self.smoothing_noise <= 0:
+                raise ValueError("Invalid value for 'smoothing_noise': %d. "
+                                 "The noise std. dev. should be greater than "
+                                 "0." % self.smoothing_noise)
+
+        # we only accept positive sparse matrix when ignore_implicit_zeros is
+        # false
+        if (not self.ignore_implicit_zeros and
+                (sparse.issparse(X) and np.any(X.data < 0))):
+            raise ValueError('QuantileTransformer only accepts non-negative'
+                             ' sparse matrices')
+
+        if sparse.issparse(X):
+            self._sparse_fit(X, rng)
+        else:
+            self._dense_fit(X, rng)
+
+        return self
+
+    def _transform_col(self, X_col, feature_idx, inverse):
+        """Private function to transform a single feature"""
+
```

Comment:
I think either do `quantiles = self.quantiles_[feature_idx]` here, or have feature quantiles passed in directly.

You might also consider having this as a function outside the class:
```python
def _quantile_transform_vector(x, quantiles, references, output_distribution, inverse):
```
dense_transform can then look something like:
```python
distrib = getattr(stats, self.output_distribution)
for x, q in zip(X.T, self.quantiles_.T):
    x[:] = _quantile_transform_vector(x, q, r, distrib, inverse=False)
```
There might be disadvantages from an efficiency perspective.